### ISSUE TITLE
parallel_localization launches one process per language, overwhelming build machines

### ISSUE THREAD

**Issue by eighthave** [untra/polyglot#78]
parallel_localization launches one process per language, in parallel, regardless of how many available cores the current machine has.  So for https://f-droid.org, we support 8 languages, and it launches 8 parallel processes on out CI/CD builds.  CI builds are virtualized, so they are rarely more than 1-2 cores.  Running 8 processes on 1 core drags the whole thing down.

Instead, it should check how many cores are available, and only schedule that many parallel processes.  Checking how many cores are available is easy to do on most platforms:
https://stackoverflow.com/questions/891537/detect-number-of-cpus-installed

**Comment by eighthave** [comment]
I should add: we were building with 11 languages, and it took 500 minutes to build.  With 8 languages, it is around 100 minutes.

**Comment by untra** [comment]
Hi @eighthave 👋 

So parallel builds will be removed in version 2.0. You can turn it off now by setting `parallel_localization: false` in your `_config.yml`. That will ensure builds are run in serial instead. Have you tried using that?

**Comment by eighthave** [comment]
yup, turning it off helped a lot.  It would be cool if it could utilize
all available cores.  Our CI systems are 1-2 core, but our deploy system
is 8 cores.  We've built a huge site, maybe 15,000 pages?  It takes an
hour to build, so speeding up the deploy would be nice.  On that
machine, parallel_localization is quite useful.

Here's the staging version of the site with polyglot:

https://fdroid.gitlab.io/fdroid-website

**Comment by untra** [comment]
Hmm... it looks like you'll want to turn the `parallel_builds` on for your production build, and turned off for your CI system.

Native Parallel Jekyll builds [was a considered feature about a year ago](https://github.com/jekyll/jekyll/pull/5344). It has since been abandoned because- well, ruby.

Polyglot parallel builds are actually just naive forks. This means it has never worked for windows machines, and really only makes sense to build almost-identical looking jekyll websites in different languages. Other complex multi-lang website features aren't possible with its design.
`parallel_builds` may help speed things up for big projects with lots of languages, but needs to go if polyglot can ever get whitelisted to run on gh-pages. 'Tis a damn shame.

I actually have stopped doing jekyll development full-time. This is now kindof a side project. But I want to get a 2.0.0 beta out soon, before 2018 preferably.

Very cool to see another big website built with polyglot! 😊 Would you mind if I updated the README with a link to your website? 🙏 

Cheers,
@untra

**Comment by eighthave** [comment]
Our site fits the description of a site that is the same except for the
language.

sure, include our site on your list :)

**Comment by vlsi** [comment]
>I should add: we were building with 11 languages, and it took 500 minutes to build

Does it really take you 10 hours to build?

I can build it in a couple of minutes (5 core, 2.6GHz Macbook).
Here's what I get:
```
bundle exec jekyll build
...
done in 121.781 seconds
```

@eighthave , what do you think?

The main time consumers for f-droid website are:
1) `fdroid-website/_plugins/git-tag-tag.rb`
It executes `git().describe(...)` again and again, and I believe git revision does not change during the build.

```ruby
        class GitTag < GitInfoTag
        include Jekyll::LiquidExtensions
                @@descr='' # <-- vlsi: I've no idea how to cache things properly, however this works
                def render(context)
                        return @@descr if @@descr != ''
                        @@descr = git().describe('HEAD', :always => true)
                end
        end

    #
    # Show the hash of HEAD.
    #
    class GitHash < GitInfoTag
        include Jekyll::LiquidExtensions
                @@sha = '' # <-- vlsi: I've no idea how to cache things properly, however this works
                def render(context)
                        return @@sha if @@sha != ''
            @@sha = git().object('HEAD').sha[0..9]
                end
    end
end
```
2) The same thing for `F-Droid's Jekyll Plugin`
What it does, it repeatedly reparses "FDroid::IndexV1", however I guess it should cache the result instead of reparsing the source JSON.
```ruby
                @@val = '' # <-- vlsi: I've no idea how to cache things properly, however this works
                def initialize(tag_name, text, tokens)
                        super
                end

                def render(context)
                        return @@val if @@val != ''
                        site = context.registers[:site]
                        url = site.config['fdroid-repo']
                        index = FDroid::IndexV1.download(url, 'en')

                        @@val = "#{index.repo.name} #{index.repo.date}"
```


Feel free to incorporate the above on t

**Comment by eighthave** [comment]
Yes, it was taking literally ~6 hours to run.  A big part of that was that its running on CI VMs that have 2gigs of RAM, and when it launched 11 parallel jobs, that requires ~4gigs of RAM.  That scenario always results in things running really slowly.

@pserwylo see above :)

**Comment by pserwylo** [comment]
Thanks @untra + @vlsi, really appreciate the insight.

This has indeed been resolved in as much as our build now takes a reasonable amount of time (e.g. 2mins on my new ThinkPad with a Core i7-7700 (4 cores with hyperthreading) + 16GiB of RAM).

Having said that, I think adding an option for "Max processes" would still be a good feature to add to the config, because that dev machine was completely unresponsive during the build. Would be great to be able to do something similar to `make --jobs=4` ([docs](https://www.gnu.org/software/make/manual/html_node/Parallel.html#Parallel)).

**Comment by Phantasus** [comment]
@eighthave I have an adhoc rough sketch, not really tested, totally expirmental.

But could that help you?: https://github.com/Phantasus/polyglot/tree/tamed_parallel_localization

**Comment by eighthave** [comment]
@Phantasus thanks for taking that on!  Looks like the right idea, we'll know better once it actually runs.  You can see a test run here:

https://gitlab.com/eighthave/fdroid-website/-/jobs/47410406

If you fork [that repo](https://gitlab.com/eighthave/fdroid-website), then you can easily change the commit in _Gemfile.lock_ and push your commits.  Then GitLab CI will automatically run it.

**Comment by eighthave** [comment]
@Phantasus I tried a batch with about 30 languages and it looks like it is working.  It got stuck after a while.  Each batch was taking about 5 minutes, then the last one hung for about 40 minutes til the CI job was killed. 
https://gitlab.com/eighthave/fdroid-website/-/jobs/48327597

Another thing I noticed is that for each batch, it says "i18n: Ensuring default English translation is in webroot".  That makes me think that is something that should only be run a single time, not while building each batch.

**Comment by eighthave** [comment]
I retried the CI job to see if it was just a temporary glitch https://gitlab.com/eighthave/fdroid-website/-/jobs/48340719

**Comment by eighthave** [comment]
The retried job got stuck in the same place.  Anything I should try next?

Also, it would be very cool if the default number of jobs per batch was derived from the CPU count in _/proc/cpuinfo_, if that is present and readable.  Here's how I do that in Python:

```python
        if os.path.exists('/proc/cpuinfo'):
            with open('/proc/cpuinfo') as fp:
                processor_core_count = len(re.findall(r'processor\s*:', fp.read()))
```

**Comment by Phantasus** [comment]
@eighthave I spotted the reason for "i18n: Ensuring default English translation is in webroot.", that's 
the fdroid-website post hook, which doesn't properly detect anymore the lockfiles for the run languages
and so gets executed. 

Sigh! interesting interconnection of hooks, side effects are fun. Let me think about that.

**Comment by Phantasus** [comment]
@eighthave how about 16 languages in 1 hour: https://gitlab.com/Phantasus/fdroid-website/-/jobs/48845881

**Comment by eighthave** [comment]
Hmm, hit the 1 hour timeout again.  It is possible to change that timer to whatever in your gitlab repo's settings, but I think things are hanging since the first few batches run pretty quickly.

**Comment by eighthave** [comment]
It might be worth setting the timeout to like 10 hours, and clicking Retry on that job.  That'll at least tell you whether the job is hung or just running really slowly.  I did that with this job:
https://gitlab.com/eighthave/fdroid-website/-/jobs/49119984

You can find that setting here:
https://gitlab.com/Phantasus/fdroid-website/settings/ci_cd

**Comment by eighthave** [comment]
Ok, after almost 6 hours it was still hung, and gave errors about running out of memory.  My guess is that for some reason when a batch is finished, it does not de-allocate all its memory.

https://gitlab.com/eighthave/fdroid-website/-/jobs/49119984

**Comment by eighthave** [comment]
I ran this with 10 languages on a machine that has 8GB RAM and 8GB swap.  It didn't run out of memory, but it did crash:

```console
$ bundle exec jekyll build -d $OUT_DIR --config _config.yml,userconfig.yml --trace
Configuration file: _config.yml
Configuration file: userconfig.yml
            Source: /builds/project-0
       Destination: build
 Incremental build: disabled. Enable with --incremental
      Generating... 
Running batch: ["bo", "en", "de", "es", "fr", "tr", "uk", "zh_Hans"]
         AutoPages: Disabled/Not configured in site.config.
         AutoPages: Disabled/Not configured in site.config.
         AutoPages: Disabled/Not configured in site.config.
         AutoPages: Disabled/Not configured in site.config.
         AutoPages: Disabled/Not configured in site.config.
         AutoPages: Disabled/Not configured in site.config.
         AutoPages: Disabled/Not configured in site.config.
         AutoPages: Disabled/Not configured in site.config.
              Lunr: Creating search index (lunr.js version 2.0.4)...
              Lunr: Creating search index (lunr.js version 2.0.4)...
              Lunr: Creating search index (lunr.js version 2.0.4)...
              Lunr: Creating search index (lunr.js version 2.0.4)...
              Lunr: Creating search index (lunr.js version 2.0.4)...
              Lunr: Index ready (lunr.js v2.0.4)
              Lunr: Added required assets to js
              Lunr: Index ready (lunr.js v2.0.4)
              Lunr: Added required assets to js
              Lunr: Index ready (lunr.js v2.0.4)
              Lunr: Added required assets to js
              Lunr: Creating search index (lunr.js version 2.0.4)...
              Lunr: Index ready (lunr.js v2.0.4)
              Lunr: Added required assets to js
              Lunr: Index ready (lunr.js v2.0.4)
              Lunr: Added required assets to js
              Lunr: Creating search index (lunr.js version 2.0.4)...
              Lunr: Creating search index (lunr.js version 2.0.4)...
        Pagination: Complete, processed 2 pagination page(s)
        Pagination: Complete, processed 2 pagination page(s)
              Lunr: Index ready (lunr.js v2.0.4)
              Lunr: Added required assets to js
        Pagination: Complete, processed 2 pagination page(s)
        Pagination: Complete, processed 2 pagination page(s)
        Pagination: Complete, processed 2 pagination page(s)
              Lunr: Index ready (lunr.js v2.0.4)
              Lunr: Added required assets to js
              Lunr: Index ready (lunr.js v2.0.4)
              Lunr: Added required assets to js
        Pagination: Complete, processed 2 pagination page(s)
        Pagination: Complete, processed 2 pagination page(s)
        Pagination: Complete, processed 2 pagination page(s)
Running batch: ["zh_Hant", "_"]
         AutoPages: Disabled/Not configured in site.config.
              Lunr: Creating search index (lunr.js version 2.0.4)...
              Lunr: Index ready (lunr.js v2.0.4)
              Lunr: Added required assets to js
        Pagination: Complete, processed 2 pagination page(s)
         AutoPages: Disabled/Not configured in site.config.
              Lunr: Creating search index (lunr.js version 2.0.4)...
              Lunr: Index ready (lunr.js v2.0.4)
              Lunr: Added required assets to js
        Pagination: Complete, processed 2 pagination page(s)
            FDroid: Can't find localized screenshot for _ at assets/fdroid-screenshot-_.png or , falling back to assets/fdroid-screenshot-en.png
Errno::ENOENT: No such file or directory @ unlink_internal - /tmp/fdroid-website._.writing.lock
  /usr/local/lib/ruby/2.3.0/fileutils.rb:1452:in `unlink'
  /usr/local/lib/ruby/2.3.0/fileutils.rb:1452:in `block in remove_file'
  /usr/local/lib/ruby/2.3.0/fileutils.rb:1457:in `platform_support'
  /usr/local/lib/ruby/2.3.0/fileutils.rb:1451:in `remove_file'
  /usr/local/lib/ruby/2.3.0/fileutils.rb:795:in `remove_file'
  /usr/local/lib/ruby/2.3.0/fileutils.rb:573:in `block in rm'
  /usr/local/lib/ruby/2.3.0/fileutils.rb:572:in `each'
  /usr/local/lib/ruby/2.3.0/fileutils.rb:572:in `rm'
  /builds/project-0/_plugins/prepare-multi-lang.rb:17:in `block in <top (required)>'
  /builds/project-0/vendor/ruby/2.3.0/gems/jekyll-3.4.3/lib/jekyll/hooks.rb:98:in `block in trigger'
  /builds/project-0/vendor/ruby/2.3.0/gems/jekyll-3.4.3/lib/jekyll/hooks.rb:97:in `each'
  /builds/project-0/vendor/ruby/2.3.0/gems/jekyll-3.4.3/lib/jekyll/hooks.rb:97:in `trigger'
  /builds/project-0/vendor/ruby/2.3.0/bundler/gems/polyglot-64b8ec212dc2/lib/jekyll/polyglot/patches/jekyll/site.rb:37:in `process'
  /builds/project-0/vendor/ruby/2.3.0/gems/jekyll-3.4.3/lib/jekyll/command.rb:26:in `process_site'
  /builds/project-0/vendor/ruby/2.3.0/gems/jekyll-3.4.3/lib/jekyll/commands/build.rb:63:in `build'
  /builds/project-0/vendor/ruby/2.3.0/gems/jekyll-3.4.3/lib/jekyll/commands/build.rb:34:in `process'
  /builds/project-0/vendor/ruby/2.3.0/gems/jekyll-3.4.3/lib/jekyll/commands/build.rb:16:in `block (2 levels) in init_with_program'
  /builds/project-0/vendor/ruby/2.3.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `block in execute'
  /builds/project-0/vendor/ruby/2.3.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `each'
  /builds/project-0/vendor/ruby/2.3.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `execute'
  /builds/project-0/vendor/ruby/2.3.0/gems/mercenary-0.3.6/lib/mercenary/program.rb:42:in `go'
  /builds/project-0/vendor/ruby/2.3.0/gems/mercenary-0.3.6/lib/mercenary.rb:19:in `program'
  /builds/project-0/vendor/ruby/2.3.0/gems/jekyll-3.4.3/exe/jekyll:13:in `<top (required)>'
  /builds/project-0/vendor/ruby/2.3.0/bin/jekyll:23:in `load'
  /builds/project-0/vendor/ruby/2.3.0/bin/jekyll:23:in `<top (required)>'
bundler: failed to load command: jekyll (/builds/project-0/vendor/ruby/2.3.0/bin/jekyll)
```

**Comment by untra** [comment]
@eighthave Holy smokes! I never imagined polyglot or jekyll would be pushed to this extreme.
I've never encountered a website so large, and with so many languages that it needed to be built for. The fdroid-website is _the_ use case for polyglot's forked builds. 👏 

Whats that `prepare-multi-lang.rb` file in the stacktrace? Is that custom?

**Comment by eighthave** [comment]
yeah, `prepare-multi-lang.rb is custom.`  Here's the repo: https://gitlab.com/fdroid/fdroid-website

This website is pretty nuts, but it works.  We had no idea that we were building the largest Jekyll site by far when we starting out.  Its about ~1500 pages per language.  We have 9 languages in production, and ~30 in staging.

**Comment by Phantasus** [comment]
@eighthave ok, I see. Next time I'll test the thing more properly. I didn't test it with a small amount so the triggers weren't executed. Sorry for wasting your time.

**Comment by eighthave** [comment]
if you get this working, that'll be huge!  Thanks what's most important :-)

**Comment by eighthave** [comment]
We got a grant to finalize the F-Droid localization, so there will soon be some funds to get this fixed.  Anyone on this thread available for a small contract?

**Comment by eighthave** [comment]
Could be like a bug bounty: fix this bug for money.

**Comment by eighthave** [comment]
`parallel_localization: false` processes the languages linearly, which would make building the site with 30 languages really slow (see https://staging.f-droid.org).  @Phantasus got a batch processing mode working, so the build can process one language per CPU core.  The problem is that the memory is not deallocated, so even though the processing is happening 4 or 8 languages at a time, the memory usage only goes up.  So this batch processing ends up the same amount of RAM as the current "all at once" mode.

**Comment by vlsi** [comment]
@eighthave , 
```
bundle exec jekyll build
...
                    done in 210.023 seconds.
```

What is the expected time to build the site?

**Comment by eighthave** [comment]
in closer to real world conditions, building 4 languages, we're
currently seeing 15-20 minutes:
https://gitlab.com/fdroid/fdroid-website/-/jobs/65365731

But the bigger issue honestly is the memory usage, since we current need
16gigs dedicated RAM to build staging.f-droid.org with all languages.
That's with 30 languages.

**Comment by tuxayo** [comment]
Hi,
I there something an outsider can do?
On the technical side. Or on the financial side.
Like putting a bounty: https://www.bountysource.com/issues/49853554-parallel_localization-launches-one-process-per-language-overwhelming-build-machines

**Comment by eighthave** [comment]
An outsider can totally solve this by reproducing the issue, then fixing
it.  Just run the whole f-droid.org build process to reproduce it, its
all free software and somewhat documented.

**Comment by eighthave** [comment]
so I've been rnuning @phantasus' batch processing change on staging.f-droid.org, which is a really large site.  The batching works, and the whole thing completes, but it has odd behavior.  Basically, the last batch takes a really long time to complete.  It looks like it stops using CPU, then spends a lot of time doing disk I/O.

**Comment by eighthave** [comment]
@obfusk since you've already done some work optimizing the Ruby code for f-droid.org, this issue might interest you.  Fixing this would allow us to deploy more languages on f-droid.org.  There are currently 3 more languages waiting to be added.

**Comment by obfusk** [comment]
This seems to work:

```diff
diff --git a/lib/jekyll/polyglot/patches/jekyll/site.rb b/lib/jekyll/polyglot/patches/jekyll/site.rb
index aa87266..8656752 100644
--- a/lib/jekyll/polyglot/patches/jekyll/site.rb
+++ b/lib/jekyll/polyglot/patches/jekyll/site.rb
@@ -1,3 +1,5 @@
+require 'etc'
+
 include Process
 module Jekyll
   class Site
@@ -32,21 +34,33 @@ module Jekyll
       prepare
       all_langs = (@languages + [@default_lang]).uniq
       if @parallel_localization
+        nproc = Etc.nprocessors
         pids = {}
-        all_langs.each do |lang|
-          pids[lang] = fork do
-            process_language lang
-          end
-        end
         Signal.trap('INT') do
           all_langs.each do |lang|
+            next unless pids.key? lang
             puts "Killing #{pids[lang]} : #{lang}"
             kill('INT', pids[lang])
           end
         end
         all_langs.each do |lang|
-          waitpid pids[lang]
-          detach pids[lang]
+          while pids.length >= nproc
+            wait
+            pids.each do |lang, pid|
+              begin
+                next unless waitpid pid, Process::WNOHANG
+              rescue SystemCallError => e
+                raise unless e.errno == Errno::ECHILD::Errno
+              end
+              pids.delete lang
+            end
+          end
+          pids[lang] = fork do
+            process_language lang
+          end
+        end
+        all_langs.each do |lang|
+          waitpid pids[lang] if pids.key? lang
         end
       else
         all_langs.each do |lang|
```

**Comment by eighthave** [comment]
I don't know much about Ruby, is `require 'etc'` part of the standard lib?  It 
does not seem to be in Debian: https://packages.debian.org/search?keywords=etc

**Comment by obfusk** [comment]
> I don't know much about Ruby, is `require 'etc'` part of the standard lib? It does not seem to be in Debian: https://packages.debian.org/search?keywords=etc

[Yes.](https://ruby-doc.org/stdlib-3.0.1/libdoc/etc/rdoc/Etc.html)

**Comment by eighthave** [comment]
if that patch was in a fork, then I can easily do a test run on staging.f-droid.org

**Comment by obfusk** [comment]
> if that patch was in a fork, then I can easily do a test run on staging.f-droid.org

See #140. That's based on `master` here though, which is a few releases ahead of https://gitlab.com/fdroid/jekyll-polyglot.

**Comment by eighthave** [comment]
I just had a quick look.  Release 1.3.3 looks like a trivial upgrade to use for 
f-droid.org.  1.4.0 and newer would need testing and review.  I haven't looked 
into what changed with the URL relativization, so I can't say what that would 
take.  How would you like to proceed?

**Comment by obfusk** [comment]
> I just had a quick look. Release 1.3.3 looks like a trivial upgrade to use for f-droid.org. 1.4.0 and newer would need testing and review. I haven't looked into what changed with the URL relativization, so I can't say what that would take. How would you like to proceed?

For now, I've cherry-picked the commit on top of 1.3.2 so you can test it in isolation: https://github.com/obfusk/polyglot/tree/nproc-1.3.2

**Comment by obfusk** [comment]
> The retried job got stuck in the same place. Anything I should try next?
> 
> Also, it would be very cool if the default number of jobs per batch was derived from the CPU count in _/proc/cpuinfo_, if that is present and readable. Here's how I do that in Python:
> 
> ```python
>         if os.path.exists('/proc/cpuinfo'):
>             with open('/proc/cpuinfo') as fp:
>                 processor_core_count = len(re.findall(r'processor\s*:', fp.read()))
> ```

FYI: in Python you can just use

```py
>>> import os
>>> len(os.sched_getaffinity(0)) # number of usable CPUs
```

or 

```py
>>> import multiprocessing
>>> multiprocessing.cpu_count() # number of CPUs in the system
```
